File: c_src/sdl_rect.c
// Copyright (c) 2014-2018, Loïc Hoguin <essen@ninenines.eu>
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice appear in all copies.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

#include "esdl2.h"

int map_to_point(ErlNifEnv* env, ERL_NIF_TERM map, SDL_Point* point)
{
	ERL_NIF_TERM x, y;

	if (!enif_get_map_value(env, map, atom_x, &x))
		return 0;
	if (!enif_get_map_value(env, map, atom_y, &y))
		return 0;

	if (!enif_get_int(env, x, &point->x))
		return 0;
	if (!enif_get_int(env, y, &point->y))
		return 0;

	return 1;
}

ERL_NIF_TERM point_to_map(ErlNifEnv* env, const SDL_Point* point)
{
	ERL_NIF_TERM map;

	map = enif_make_new_map(env);

	enif_make_map_put(env, map, atom_x,
		enif_make_int(env, point->x), &map);
	enif_make_map_put(env, map, atom_y,
		enif_make_int(env, point->y), &map);

	return map;
}

int map_to_rect(ErlNifEnv* env, ERL_NIF_TERM map, SDL_Rect* rect)
{
	ERL_NIF_TERM x, y, w, h;

	if (!enif_get_map_value(env, map, atom_x, &x))
		return 0;
	if (!enif_get_map_value(env, map, atom_y, &y))
		return 0;
	if (!enif_get_map_value(env, map, atom_w, &w))
		return 0;
	if (!enif_get_map_value(env, map, atom_h, &h))
		return 0;

	if (!enif_get_int(env, x, &rect->x))
		return 0;
	if (!enif_get_int(env, y, &rect->y))
		return 0;
	if (!enif_get_int(env, w, &rect->w))
		return 0;
	if (!enif_get_int(env, h, &rect->h))
		return 0;

	return 1;
}

ERL_NIF_TERM rect_to_map(ErlNifEnv* env, const SDL_Rect* rect)
{
	ERL_NIF_TERM map;

	map = enif_make_new_map(env);

	enif_make_map_put(env, map, atom_x,
		enif_make_int(env, rect->x), &map);
	enif_make_map_put(env, map, atom_y,
		enif_make_int(env, rect->y), &map);
	enif_make_map_put(env, map, atom_w,
		enif_make_int(env, rect->w), &map);
	enif_make_map_put(env, map, atom_h,
		enif_make_int(env, rect->h), &map);

	return map;
}

// enclose_points

NIF_FUNCTION(enclose_points)
{
	SDL_Point* points = NULL;
	SDL_Rect clip, result;
	SDL_bool b;
	ERL_NIF_TERM list, head;
	unsigned int count;
	int i = 0;

	BADARG_IF(!enif_get_list_length(env, argv[0], &count));
	BADARG_IF(!map_to_rect(env, argv[1], &clip));

	points = (SDL_Point*)enif_alloc(count * sizeof(SDL_Point));

	list = argv[0];
	while (enif_get_list_cell(env, list, &head, &list)) {
		if (!map_to_point(env, head, &(points[i++]))) {
			enif_free(points);
			return enif_make_badarg(env);
		}
	}

	b = SDL_EnclosePoints(points, count, &clip, &result);

	enif_free(points);

	if (!b)
		return atom_false;

	return rect_to_map(env, &result);
}

// has_intersection

NIF_FUNCTION(has_intersection)
{
	SDL_Rect a, b;

	BADARG_IF(!map_to_rect(env, argv[0], &a));
	BADARG_IF(!map_to_rect(env, argv[1], &b));

	if (SDL_HasIntersection(&a, &b))
		return atom_true;

	return atom_false;
}

// intersect_rect

NIF_FUNCTION(intersect_rect)
{
	SDL_Rect a, b, result;

	BADARG_IF(!map_to_rect(env, argv[0], &a));
	BADARG_IF(!map_to_rect(env, argv[1], &b));

	if (!SDL_IntersectRect(&a, &b, &result))
		return atom_false;

	return rect_to_map(env, &result);
}

// intersect_rect_and_line

NIF_FUNCTION(intersect_rect_and_line)
{
	SDL_Rect rect;
	int x1, y1, x2, y2;

	BADARG_IF(!map_to_rect(env, argv[0], &rect));
	BADARG_IF(!enif_get_int(env, argv[1], &x1));
	BADARG_IF(!enif_get_int(env, argv[2], &y1));
	BADARG_IF(!enif_get_int(env, argv[3], &x2));
	BADARG_IF(!enif_get_int(env, argv[4], &y2));

	if (!SDL_IntersectRectAndLine(&rect, &x1, &y1, &x2, &y2))
		return atom_false;

	return enif_make_tuple4(env,
		enif_make_int(env, x1),
		enif_make_int(env, y1),
		enif_make_int(env, x2),
		enif_make_int(env, y2));
}

// union_rect

NIF_FUNCTION(union_rect)
{
	SDL_Rect a, b, result;

	BADARG_IF(!map_to_rect(env, argv[0], &a));
	BADARG_IF(!map_to_rect(env, argv[1], &b));

	SDL_UnionRect(&a, &b, &result);

	return rect_to_map(env, &result);
}

File: src/sdl_rect.erl
%% Copyright (c) 2014-2018, Loïc Hoguin <essen@ninenines.eu>
%%
%% Permission to use, copy, modify, and/or distribute this software for any
%% purpose with or without fee is hereby granted, provided that the above
%% copyright notice and this permission notice appear in all copies.
%%
%% THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
%% WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
%% MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
%% ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
%% WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
%% ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
%% OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

%% The SDL_FORCE_INLINE functions have been reimplemented in Erlang.
-module(sdl_rect).

-export([enclose_points/2]).
-export([has_intersection/2]).
-export([intersection/2]).
-export([intersection/3]).
-export([intersection/5]).
-export([is_empty/1]).
-export([is_equal/2]).
-export([is_in/2]).
-export([union/2]).

-type point() :: #{x=>integer(), y=>integer()}.
-export_type([point/0]).

-type rect() :: #{x=>integer(), y=>integer(), w=>integer(), h=>integer()}.
-export_type([rect/0]).

-spec enclose_points([point()], rect()) -> rect() | false.
enclose_points(Points, ClipRect) ->
	esdl2:enclose_points(Points, ClipRect).

-spec has_intersection(rect(), rect()) -> boolean().
has_intersection(Rect1, Rect2) ->
	esdl2:has_intersection(Rect1, Rect2).

-spec intersection(rect(), rect()) -> rect() | false.
intersection(Rect1, Rect2) ->
	esdl2:intersect_rect(Rect1, Rect2).

-spec intersection(rect(), P1, P2) -> {P1, P2} | false when P1 :: point(), P2 :: point().
intersection(Rect, #{x := X1, y := Y1}, #{x := X2, y := Y2}) ->
	case intersection(Rect, X1, Y1, X2, Y2) of
		false ->
			false;
		{ResX1, ResY1, ResX2, ResY2} ->
			{#{x => ResX1, y => ResY1}, #{x => ResX2, y => ResY2}}
	end.

-spec intersection(rect(), X1, Y1, X2, Y2) -> {X1, Y1, X2, Y2} | false
	when X1 :: integer(), Y1 :: integer(), X2 :: integer(), Y2 :: integer().
intersection(Rect, X1, Y1, X2, Y2) ->
	esdl2:intersect_rect_and_line(Rect, X1, Y1, X2, Y2).

-spec is_empty(rect()) -> boolean().
is_empty(#{w := W, h := H}) ->
	(W =< 0) orelse (H =< 0).

-spec is_equal(rect(), rect()) -> boolean().
is_equal(Rect1, Rect2) ->
	Rect1 =:= Rect2.

-spec is_in(point(), rect()) -> boolean().
is_in(#{x := PX, y := PY}, #{x := RX, y := RY, w := RW, h := RH}) ->
	(PX >= RX) andalso (PX < RX + RW)
		andalso (PY >= RY) andalso (PY < RY + RH).

-spec union(rect(), rect()) -> rect().
union(Rect1, Rect2) ->
	esdl2:union_rect(Rect1, Rect2).
